### PR TITLE
Fix issues with using block param named `component`.

### LINKED ISSUE #14096: Using `component` as a yielded block param leads to bizarre error condition.
Using the word `component` as a yielded block param leads to a very odd error message in glimmer2.

Sample invocation:

```hbs
{{#foo-bar as |component|}}
  {{component.title}}
{{/foo-bar}}
```

Error:

```
Error: Assertion Failed: You cannot create a component from null using the {{component}} helper
    at assert (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:5682:13)
    at Object.assert (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:14586:34)
    at ClosureComponentReference.compute (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:8855:26)
    at ClosureComponentReference.value (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:12095:45)
    at GuardedCautiousAppendOpcode.evaluate (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:43893:51)
    at VM.execute (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:49785:28)
    at Template.render (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:49318:23)
    at render (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:10093:46)
    at runInTransaction (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:20230:9)
    at transaction (https://s3.amazonaws.com/builds.emberjs.com/alpha/ember.debug.js:10115:27)
```

[Demo JSBin](https://jsbin.com/tavofu/edit?html,js,output)

### PR DESCRIPTION
**Pull request by rwjblue**
Fixes #14096.

### CODE REVIEW

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js [hunk 1]
```diff
@@ -602,6 +602,26 @@ moduleFor('Components test: curly components', class extends RenderingTest {
...
+        this._super(...arguments);
+        instance = this;
+      },
+
+      name: 'foo-bar'
+    });
+
+    this.registerComponent('foo-bar', { ComponentClass: FooBarComponent, template: '{{yield this}}' });
+
+    debugger
```
Comment (kiwiupover) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: Do you have a test failing in CI that passes locally? :P

Comment (rwjblue) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: haha, no I just forgot it

**Comment by rwjblue** [comment]
Updated with some fixes and a bit of refactoring around the helper lookup system.  Now passing and ready for review.

File: packages/ember-glimmer/lib/environment.js [hunk 2]
```diff
@@ -215,19 +213,14 @@ export default class Environment extends GlimmerEnvironment {
...
-        return DynamicComponentSyntax.create({ args, templates, symbolTable });
-      } else if (key === 'render') {
-        return new RenderSyntax({ args });
-      } else if (key === 'outlet') {
-        return new OutletSyntax({ args });
-      } else if (key === 'mount') {
-        return new MountSyntax({ args });
+      let buildRefinedSyntax = findSyntaxBuilder(key);
+      if (buildRefinedSyntax) {
+        return buildRefinedSyntax({ args, templates, symbolTable });
```
Comment (chancancode) [inline packages/ember-glimmer/lib/environment.js]: While you are at it, can you make these positional arguments? I started doing it in https://github.com/emberjs/ember.js/pull/14172 so I think we need to pass a common set of args with your approach. `this, args, templates, symbolTable` is probably okay.

Comment (rwjblue) [inline packages/ember-glimmer/lib/environment.js]: 👍  - Will update.

File: packages/ember-glimmer/tests/integration/helpers/custom-helper-test.js [hunk 3]
```diff
@@ -27,6 +27,18 @@ moduleFor('Helpers test: custom helpers', class extends RenderingTest {
...
+    this.registerHelper('hello.world', () => 'hello world');
+
+    this.render('{{hello.world}}{{hello.world foo="bar"}}');
+
+    this.assertText('');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('');
+  }
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/helpers/custom-helper-test.js]: Should probably check that it actually worked (by setting "context.hello.world", doing the I-N-U-R)

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js [hunk 4]
```diff
@@ -602,6 +602,25 @@ moduleFor('Components test: curly components', class extends RenderingTest {
...
+
+      name: 'foo-bar'
+    });
+
+    this.registerComponent('foo-bar', { ComponentClass: FooBarComponent, template: '{{yield this}}' });
+
+    this.render('{{#foo-bar as |component|}}{{component.name}}{{/foo-bar}}');
+
+    this.assertComponentElement(this.firstChild, { content: 'foo-bar' });
+  }
```
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: missing -N-U-R

File: packages/ember-glimmer/lib/syntax.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+registerSyntax('outlet', function(options) {
+  return new OutletSyntax(options);
+});
+
+registerSyntax('mount', function(options) {
+  return new MountSyntax(options);
+});
+
+registerSyntax('component', function(options) {
+  return DynamicComponentSyntax.create(options);
```
Comment (chancancode) [inline packages/ember-glimmer/lib/syntax.js]: We can probably just make all syntax have a `create` that takes care of arguments normalization, assertions etc, like in https://github.com/emberjs/ember.js/pull/14172.

Then we can just do `registerSyntax('component', DynamicComponentSyntax)`, etc, and call `.create` on the other side.

**Comment by rwjblue** [comment]
Fixed the issues with the tests, removed argument destructuring, and rebased.

**Pull request merged**